### ISSUE TITLE
Add support for structured logs

### ISSUE THREAD

**Issue by TrevinTeacutter** [projectcontour/contour#624]
It would be nice to have access logs that are structured as we use splunk for our logging aggregation and extracting key/value pairs is much nicer in splunk with structured logs than it is with regex. Could this be some sort of flag to enable this or did I just miss that this option is already enabled somewhere?

Blocked
- [x] #1130

**Comment by davecheney** [comment]
Thank you for raising this issue. The access logs are generated directly from envoy. At the moment we have flags to control their output destination, but their format is currently defaulting to whatever envoy uses by default -- something akin to apache2 log.

https://github.com/heptio/contour/blob/master/internal/contour/listener.go#L338

Looking at 

https://www.envoyproxy.io/docs/envoy/v1.6.0/configuration/access_log#config-access-log-format

It looks like its possible to redefine the output of the logging format.

I'm wary of exposing the full configuration syntax of this to end users, it's too close to showing how the sausage is made. With that said, when you say _structured logging_, do you mean JSON?

**Comment by TrevinTeacutter** [comment]
Yes, I mean JSON. We recently turned JSON access logs on for Traefik in our old DC/OS clusters and they have been useful as well as capable of including/redacting/excluding headers of requests coming through the ingress making correlation really easy, stuff that normally doesn't get included in apache2 logs for brevity's sake.

**Comment by TrevinTeacutter** [comment]
Yes, I mean JSON. We recently turned JSON access logs on for Traefik in our old DC/OS clusters and they have been useful as well as capable of including/redacting/excluding headers of requests coming through the ingress making correlation really easy, stuff that normally doesn't get included in apache2 logs for brevity's sake.

**Comment by 01deyishu** [comment]
Maybe support define log format by configmap is a good way , just like nginx-ingress-controller use log-format-upstream parameter, also istio use logentry is a good practice.

**Comment by davecheney** [comment]
We're not going to get to this for the 0.7 release. Removing milestone.

**Comment by azuretek** [comment]
Has there been any agreement on this? Right now this is making Contour a less usable replacement for Nginx, having Envoy metrics available is great but getting full access log data would go a long way to making this a viable option.

**Comment by davecheney** [comment]
No there has been no agreement because, apart from a request for json formatted logs, no details have been provided.

**Comment by azuretek** [comment]
For our use case it would be nice if we could add an annotation or even just load a ConfigMap that will add a structure like so to every listener filter chain.

```
access_log: [
  {
    name: "envoy.file_access_log",
    config: {
      json_format: {
        downstream_remote_address: "%DOWNSTREAM_REMOTE_ADDRESS%",
        path: "%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%",
        authority: "%REQ(:AUTHORITY)%",
        protocol: "%PROTOCOL%",
        upstream_service_time: "%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%",
        upstream_local_address: "%UPSTREAM_LOCAL_ADDRESS%",
        duration: "%DURATION%",
        downstream_local_address: "%DOWNSTREAM_LOCAL_ADDRESS%",
        user_agent: "%REQ(USER-AGENT)%",
        response_code: "%RESPONSE_CODE%",
        response_flags: "%RESPONSE_FLAGS%",
        start_time: "%START_TIME%",
        method: "%START_TIME%",
        request_id: "%REQ(X-REQUEST-ID)%",
        upstream_host: "%UPSTREAM_HOST%",
        x_forwarded_for: "%REQ(X-FORWARDED-FOR)%",
        requested_server_name: "%REQUESTED_SERVER_NAME%",
        bytes_received: "%BYTES_RECEIVED%",
        bytes_sent: "%BYTES_SENT%",
        upstream_cluster: "%UPSTREAM_CLUSTER%"
      },
      path: "/dev/stdout"
    },
  },
]
```

It looks like we can probably add it into this function with minimal effort https://github.com/heptio/contour/blob/master/internal/envoy/listener.go#L189

**Comment by davecheney** [comment]
Our general preference is rather than make things configurable, we choose a reasonable default for all users. Is there a standard for json logs in this format? If so, we might just make that the default for everyone.

**Comment by azuretek** [comment]
I agree with that approach, there doesn't seem to be a standard for JSON access logs but this structure appears to have everything that Envoy reports (other than custom headers)

```
json_format: {
  downstream_remote_address: "%DOWNSTREAM_REMOTE_ADDRESS%",
  path: "%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%",
  authority: "%REQ(:AUTHORITY)%",
  protocol: "%PROTOCOL%",
  upstream_service_time: "%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%",
  upstream_local_address: "%UPSTREAM_LOCAL_ADDRESS%",
  duration: "%DURATION%",
  downstream_local_address: "%DOWNSTREAM_LOCAL_ADDRESS%",
  user_agent: "%REQ(USER-AGENT)%",
  response_code: "%RESPONSE_CODE%",
  response_flags: "%RESPONSE_FLAGS%",
  start_time: "%START_TIME%",
  method: "%REQ(:METHOD)%",
  request_id: "%REQ(X-REQUEST-ID)%",
  upstream_host: "%UPSTREAM_HOST%",
  x_forwarded_for: "%REQ(X-FORWARDED-FOR)%",
  requested_server_name: "%REQUESTED_SERVER_NAME%",
  bytes_received: "%BYTES_RECEIVED%",
  bytes_sent: "%BYTES_SENT%",
  upstream_cluster: "%UPSTREAM_CLUSTER%"
}
```

**Comment by azuretek** [comment]
So I went ahead and did a build against the 0.11.0 code base with the changes we discussed.

https://github.com/heptio/contour/pull/1053

The dynamic listener logs now look like this:

```
{"upstream_cluster":"default/hello-world/8080/da39a3ee5e","downstream_remote_address":"1.2.3.4:49286","path":"/?abc=123","authority":"hello.test.foo.bar","protocol":"HTTP/2","upstream_service_time":"1","@timestamp":"2019-04-25T06:31:07.162Z","upstream_local_address":"-","duration":"1","downstream_local_address":"10.220.3.107:443","response_code":"200","user_agent":"curl/7.54.0","response_flags":"-","method":"GET","request_id":"b4ecadf1-8fc0-4e87-89d8-a3aa851196d4","upstream_host":"10.141.181.15:8080","x_forwarded_for":"1.2.3.4","requested_server_name":"hello.test.foo.bar","bytes_received":"0","uber_trace_id":"-","bytes_sent":"17"}
```

I also added the same format to the static listener configs, this is the log for the health check endpoiont:

```
{"response_code":"200","user_agent":"kube-probe/1.11","response_flags":"-","method":"GET","request_id":"aadded21-6b3c-4fa4-bc3a-de62b6b4c2ec","upstream_host":"-","x_forwarded_for":"-","requested_server_name":"-","bytes_received":"0","uber_trace_id":"-","bytes_sent":"0","upstream_cluster":"-","downstream_remote_address":"10.220.40.142:43216","path":"/healthz","authority":"10.141.207.73:8002","protocol":"HTTP/1.1","upstream_service_time":"-","@timestamp":"2019-04-25T06:31:09.869Z","upstream_local_address":"-","duration":"0","downstream_local_address":"10.141.207.73:8002"}
```

**Comment by bgagnon** [comment]
@azuretek is there a reason or restriction for outputting numerical values as strings in the JSON objects?

Bytes and duration are things one would want to do aggregations and calculate metrics on (assuming the logs end up in Elasticsearch or similar database). For the response code, it's convenient for range queries (<100, 200-299, 300-399, 400-499,>500).

It would be nice to avoid an ETL for such simple JSON type coercions.

**Comment by azuretek** [comment]
@bgagnon definitely ran into that in my PoC, it's annoying but ultimately it's able to be handled by adjusting the mappings on the elastic search side. To make it report correctly we'd have to make changes to envoy (which should probably be done ultimately).

**Comment by davecheney** [comment]
Abi,

Can you please detail the changes you require. If you can give complete examples of the expected json output that would be best. 

Thanks

Dave

**Comment by azuretek** [comment]
@davecheney I need the default access log format to be structured, preferably JSON to more easily ingest into storage backends like Elasticsearch. This is a requirement for us to move away from using NGINX as our ingress.

Currently access logs use the default format string, as defined in the envoy documentation: https://www.envoyproxy.io/docs/envoy/latest/configuration/access_log#default-format-string
```
[%START_TIME%] "%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%"
%RESPONSE_CODE% %RESPONSE_FLAGS% %BYTES_RECEIVED% %BYTES_SENT% %DURATION%
%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% "%REQ(X-FORWARDED-FOR)%" "%REQ(USER-AGENT)%"
"%REQ(X-REQUEST-ID)%" "%REQ(:AUTHORITY)%" "%UPSTREAM_HOST%"\n
```

The desired format is JSON specifically with these keys and values. As defined in the envoy documentation https://www.envoyproxy.io/docs/envoy/latest/configuration/access_log#format-dictionaries
```json
{
  "@timestamp": "%START_TIME%",
  "authority": "%REQ(:AUTHORITY)%",
  "bytes_received": "%BYTES_RECEIVED%",
  "bytes_sent": "%BYTES_SENT%",
  "downstream_local_address": "%DOWNSTREAM_LOCAL_ADDRESS%",
  "downstream_remote_address": "%DOWNSTREAM_REMOTE_ADDRESS%",
  "duration": "%DURATION%",
  "method": "%REQ(:METHOD)%",
  "path": "%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%",
  "protocol": "%PROTOCOL%",
  "request_id": "%REQ(X-REQUEST-ID)%",
  "requested_server_name": "%REQUESTED_SERVER_NAME%",
  "response_code": "%RESPONSE_CODE%",
  "response_flags": "%RESPONSE_FLAGS%",
  "uber_trace_id": "%REQ(UBER-TRACE-ID)%",
  "upstream_cluster": "%UPSTREAM_CLUSTER%",
  "upstream_host": "%UPSTREAM_HOST%",
  "upstream_local_address": "%UPSTREAM_LOCAL_ADDRESS%",
  "upstream_service_time": "%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%",
  "user_agent": "%REQ(USER-AGENT)%",
  "x_forwarded_for": "%REQ(X-FORWARDED-FOR)%"
},
```

**Comment by davecheney** [comment]
Thank you for your reply. 

My goal is to change the log output format once, to something structured, without the option to change it. 

To do this I need your assistance to know if there is an established standard for json structured http logs to replace the current ncsa style log format.

**Comment by davecheney** [comment]
Thank you to everyone who has contributed to this issue over the last few weeks.

I want to make it clear to all readers that I am not opposed to adding this feature. In fact I want to see it adopted as default. 

What I do not want to see happen is JSON or NCSA become selectable especially if there are options to make each of those formats configurable. If this is the outcome we will stick with the default NCSA format.

Please do not send any PRs until there is agreement on which format will be used.

**Comment by azuretek** [comment]
There is no JSON standard that I'm aware of, I haven't been able to find a JSON standard proposed by any other web proxy projects or otherwise. I'm not certain what additional information you need, can you be more clear about what you're looking for specifically?

If there is no standard other than NCSA you're saying we want to stick to that standard?

**Comment by moderation** [comment]
One initiative to consider for a standardized JSON format is https://cloudevents.io/. This is a working group in the CNCF looking at standardizing event data. Version 0.2 of the specification is at https://github.com/cloudevents/spec#cloudevents-documents

**Comment by azuretek** [comment]
Unfortunately Envoy doesn't support nesting in their access log output, so implementing a standard that requires nesting isn't possible. Not only that but the cloud events standard doesn't say anything about the fields and format of the data being represented, it's just adding a standard metadata to "events".

**Comment by davecheney** [comment]
I’m sorry I was not clear. My desire is to make one change to contour for all users, if that is to switch from NCSA to an as yet undefined JSON logging format then we must establish that format here in this issue. Making the configuration endlessly configurable is a support load that I cannot foist on the CRE team or future maintainers.

**Comment by azuretek** [comment]
Unless I'm unclear on what's expected I think I've proposed a format, is there a specific proposal style I should be following? I'm all for specifying a one time format as to not make it a maintenance burden, but I need some more direction to satisfy your requirements.

**Comment by davecheney** [comment]
Thank you for proposing #1080. I have several concerns which I have outlined in the review. The overriding one is the format of log lines is not a product differentiator. Sure, its annoying to integrate Contour with the in house log aggregation solution, but fundamentally once that is done, by whatever reason, this is not a place where Contour can differentiate itself above the other ingress controllers.

To that end, I want to limit the amount of sustaining engineering that is spent on this. It is clear from #1080 that we cannot make structured JSON logs the default for all users -- they are too verbose compared to the current, albeit unsatisfactory, situation.

I want to attack this problem by asking what is the best outcome. I feel the end result must be lowering the maintenance cost of ingesting Envoy's access logs. The current situation with an effectively custom output format has a cost that log aggregation services cannot currently ingest it without being taught how to interpret it. JSON logging is _a_ solution, but I want to propose another, which is Envoy should output in [CLF](https://en.wikipedia.org/wiki/Common_Log_Format). Unlike JSON this _is_ a defacto standard, and tools support it natively. 

Regardless of this discussion I think we should change Envoy's logging format to CLF, see #1124. Whether we offer other logging formats as an option remains to be seen, by my hope is that by changing Envoy to emit something which is a known standard by default, the demand for an alternative format can be reduced, or at least the experience will clarify precisely what is missing from CLF formatted logs.

**Comment by tyrken** [comment]
You see the format as not a differentiator, but for other ingresses can have fully flexible output via https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#log-format-upstream, https://istio.io/docs/tasks/telemetry/logs/access-log/ and https://github.com/jcmoraisjr/haproxy-ingress#log-format, while I suspect most other ingresses, e.g. Traefik can be globally configured to a degree (if Traefik is a little limited in request header logging).

So I'd say the lack of it is a current Contour weakness, though I admit this is seemingly shared with Ambassador & Gloo from a quick read...

**Comment by bgagnon** [comment]
> It is clear from #1080 that we cannot make structured JSON logs the default for all users

I agree; I'm not sure where it was suggested JSON logs should be on by default. They should be optional.

> I want to propose another, which is Envoy should output in CLF. Unlike JSON this is a defacto standard, and tools support it natively.

"Supporting natively" is somewhat of a stretch. A common stack like ELK will first parse CLF to JSON, then store it and index it as JSON, and finally run queries on it.

Using unstructured logging conventions like CLF dodges the issue of having to name the fields, but overall it just pushes the burden to the user, while discarding a lot of useful data in the process.

It's true that there's a lack of a standards for JSON web server logs, but there are efforts like the Elastic Common Schema to improve that:
https://www.elastic.co/guide/en/ecs/current/ecs-http.html

I don't feel this lack of standards should be an argument for not even _trying_ to name things. Since Envoy has picked some names, it's also not unreasonable to use those as @azuretek proposes.

> I feel the end result must be lowering the maintenance cost of ingesting Envoy's access logs.

I agree here. With the typical ELK stack, having JSON logs has the lowest maintenance cost because I don't need parsers (either pre-made to parse CLF or arcane regexes) in Fluent, Logstash, Filebeat or anything like that. I can simply tail the envoy logs and push the verbatim lines to Elasticsearch.

But more importantly, like @tyrken, I do see the potential for a _product differentiator_, as the logs can be enriched with Contour-specific concepts.

For example, I might favour Contour over other solutions for its multi-tenancy and delegation capabilities. It would be only natural for the access logs (which are now mixed together for multiple teams / delegation entities) to include fields for grouping and filtering the logs specific to one IngressRoute (its name, namespace, etc.) or a portion of the graph. The CLF would require a lot of post-processing to make this possible.

In @azuretek's example above, `upstream_cluster` is an example of a product-differentiating field -- it would not be visible if the log was dumbed down to CLF, and it contains useful topology that is intrinsic to Contour.

**Comment by davecheney** [comment]
@bgagnon thanks for your reply

> I agree; I'm not sure where it was suggested JSON logs should be on by default. They should be optional.

Any fool can make a product with a million configurable options. My goal for Contour is we make opinionated decisions that make the product easier to use not just _by default_, but without them having to think about it. For example, we configure HTTP/2 out of the box, you don't have to opt in. We configure gzip for every connection, so that you don't launch your service and discover that you didn't enable compression. Along this line I want there to be one way of emitting logs, currently its Envoy's custom format, ideally the resolution to this ticket would be to emit JSON formatted logs only. In this case it is not going to be possible to make this decision unilaterally.

> Using unstructured logging conventions like CLF dodges the issue of having to name the fields, but overall it just pushes the burden to the user, while discarding a lot of useful data in the process.

I agree. CLF is not perfect, but it is better than what Envoy currently emits. This is something I should have done a long time ago.

When I wrote my comment a few days ago I wanted to write a sizeable response because the OP had gone to the effort to write a proposal. In my efforts I buried the lede, so I'll restate it here for clarity.

_I am not say that we will not add JSON logging, but we will not add it yet._

At a minimum there needs to be a place to hold the configuration for the JSON format string because everyone is going to want to change it. I don't think it is reasonable to pack that into a cli flag, so I raised #1130 to track that work. 

> But more importantly, like @tyrken, I do see the potential for a product differentiator, as the logs can be enriched with Contour-specific concepts.

Could you explain more? These are the kinds of user stories I want to capture.

**Comment by davecheney** [comment]
@tyrken thank you for your reply. When I say that logging is not a product differentiator what I mean is logging is like brushing your teeth or paying your taxes, a necessary evil. Product differentiators are things that let people deploy complicated web applications on a k8s cluster using contour.

**Comment by azuretek** [comment]
@davecheney I don't think moving to CLF will solve any existing problem, as explained by others any non-structured format will still require pre-processing in some way. But also we do lose envoy specific fields like `upstream_cluster`.

I'm totally on board with scrapping this in favor of allowing dynamic config from either a configmap or some other kind of volume mounted config. The only concern is that dynamic listeners don't get their config from a config file, so we'd have to put together some way to override/inject custom config at the listener level.

**Comment by davecheney** [comment]
I’m not sure what you mean by dynamic listeners. Contour configures up to three itself, stats-listener, ingress_http and/or ingress_https. Contour controls the access log configuration for all of those listeners.

**Comment by azuretek** [comment]
Contour configures the `static_resources` via bootstrap, which is loaded via `envoy --config-path`. Whereas the upstream clusters are configured dynamically via GRPC. That's what I mean by dynamic listeners.

**Comment by davecheney** [comment]
Thank you for explaining. Bootstrap config points envoy to contour which provides it with the rest of its configuration. That configuration is either invented inside contour or translated from the contents of the k8s api server.

**Comment by bgagnon** [comment]
@davecheney Sure, happy to explain!

What I meant by enriching access logs with Contour specific concepts is having something that will reveal relevant pieces of the DAG topology in the access logs.

For starters, each log entry could contain (I am making these field names up BTW):

* `contour.ingressroute.namespace`
* `contour.ingressroute.name`
* `contour.ingressroute.fqdn`
* `contour.route.path`
* `contour.route.service.namespace`
* `contour.route.service.name`
* `contour.route.service.port`

Using this metadata, it is much easier to deal with the multiplexed logs of a multi-tenant cluster. This, I believe to be a setup where Contour shines.

From a team perspective, it is now possible:
* get all access logs for services within my namespace
* get Envoy-reported 5xx statuses for service foo in namespace bar

Compared to the undifferentiated mess of CLF, it seems a lot saner.

Once challenge I see is that while Envoy is producing the access logs, it does not really access to the Contour information unless we inject a lot more data through the discovery services.

For this reason, I am starting to think the Access Log Service (ALS) might be the superior option to achieve logging bliss: https://www.envoyproxy.io/docs/envoy/latest/api-v2/service/accesslog/v2/als.proto

Using this tap-like gRPC sink from Envoy, Contour could join the incoming log entries with the information it knows and produce the final JSON output. This would fix a lot of the shortcomings with the built-in JSON output from Envoy (typing, field nesting, etc.).

On a related note, this might be slipping into OpenTracing support (issue #399), since the fields I describe are typically those you'd want as labels/tags on your trace spans.

**Comment by davecheney** [comment]
We will add the plumbing to support this in 0.14. Marking tentatively for 0.14 but may bump to 0.15.

**Comment by davecheney** [comment]
#1130 landed in 0.14. Moving to 0.15 to implement.

**Comment by davecheney** [comment]
Sorry, we ran out of time in 0.15 for this. @youngnick has a plan for how to implement this in beta1.

**Comment by youngnick** [comment]
I've published #1485 with our current design ideas around this feature. I'm hoping to get this reviewed and agreed soon with the aim to land this in beta1.
